src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java

```diff
@@ -256,90 +267,120 @@ private static boolean shouldBeDeclaredAsFinal(ClassDesc desc) {
+                                                    qualifiedClassName, className);
+                                            });
+            nearest.ifPresentOrElse(ClassDesc::registerNestedSubclass,
+                                    () -> {
+                                        Optional
+                                            .ofNullable(innerClasses.get(superClassName))
+                                            .ifPresent(ClassDesc::registerNestedSubclass);
+                                    });
         }
     }
 
     /**
-     * Checks if there is a class with same name.
+     * Register to the super class of anonymous inner class that the given class is instantiated
+     * by an anonymous inner class.
+     *
+     * @param literalNewAst ast node of {@link TokenTypes#LITERAL_NEW} representing anonymous inner
+     *                      class
+     * @param outerTypeDeclName Fully qualified name of the outer type declaration of anonymous
+     *                          inner class
+     */
+    private void registerAnonymousInnerClassToSuperClass(DetailAST literalNewAst,
+                                                         String outerTypeDeclName) {
+        final String superClassName = CheckUtil.getShortNameOfAnonInnerClass(literalNewAst);
+        final Optional<ClassDesc> superClassDesc =
+            getNearestClassWithSameName(superClassName,
+                                        className -> {
+                                            return CheckUtil
+                                                .getAnonSuperTypeDeclarationMatchingCount(
+                                                    outerTypeDeclName, className);
+                                        });
+        superClassDesc.ifPresentOrElse(ClassDesc::registerSuperClassOfAnonymousInnerClass,
+                                       () -> {
+                                           Optional.ofNullable(innerClasses.get(superClassName))
+                                               .ifPresent(
+                                                   ClassDesc
+                                                       ::registerSuperClassOfAnonymousInnerClass);
+                                       });
+    }
+
+    /**
+     * Get the nearest class with same name.
+     *
+     * <p>The parameter {@code countProvider} exists because if the class being searched is the
+     * super class of anonymous inner class, the rules of evaluation are a bit different,
+     * consider the following example-
+     * <pre>
+     * {@code
+     * public class Main {
+     *     static class a {
+     *         static class b {
+     *         }
+     *     }
+     *
+     *     class j {
+     *         a.b object = new a.b() { // Object of Main.j.a.b and not of Main.a.b
+     *         };
+     *
+     *         static class a {
+     *             static class b {
```

Comment:
although it is just an example, please use conventional naming for classes - camelCase

Reply:
Done.